File: data/src/main/java/org/cryptomator/data/util/TransferredBytesAwareInputStream.java
package org.cryptomator.data.util;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;

public abstract class TransferredBytesAwareInputStream extends InputStream {

	private static final int EOF = -1;

	private final InputStream in;

	private long transferred;

	public TransferredBytesAwareInputStream(InputStream in) {
		this.in = in;
	}

	@Override
	public int read() throws IOException {
		int result = in.read();
		if (result != EOF) {
			bytesTransferred(++transferred);
		}
		return result;
	}

	@Override
	public int read(byte @NotNull [] b) throws IOException {
		int result = in.read(b);
		if (result != EOF) {
			transferred += result;
			bytesTransferred(transferred);
		}
		return result;
	}

	@Override
	public int read(byte @NotNull [] b, int off, int len) throws IOException {
		int result = in.read(b, off, len);
		if (result != EOF) {
			transferred += result;
			bytesTransferred(transferred);
		}
		return result;
	}

	@Override
	public void close() throws IOException {
		in.close();
	}

	@Override
	public int available() throws IOException {
		return in.available();
	}

	@Override
	public long skip(long n) throws IOException {
		long result = in.skip(n);
		transferred += result;
		bytesTransferred(transferred);
		return result;
	}

	public abstract void bytesTransferred(long transferred);
}

File: data/src/main/java/org/cryptomator/data/util/TransferredBytesAwareInputStream.kt
package org.cryptomator.data.util

import java.io.IOException
import java.io.InputStream

abstract class TransferredBytesAwareInputStream(private val inputStream: InputStream) : InputStream() {

	private var transferred: Long = 0

	@Throws(IOException::class)
	override fun read(): Int {
		val result = inputStream.read()
		if (result != EOF) {
			bytesTransferred(++transferred)
		}
		return result
	}

	@Throws(IOException::class)
	override fun read(b: ByteArray): Int {
		val result = inputStream.read(b)
		if (result != EOF) {
			transferred += result.toLong()
			bytesTransferred(transferred)
		}
		return result
	}

	@Throws(IOException::class)
	override fun read(b: ByteArray, off: Int, len: Int): Int {
		val result = inputStream.read(b, off, len)
		if (result != EOF) {
			transferred += result.toLong()
			bytesTransferred(transferred)
		}
		return result
	}

	@Throws(IOException::class)
	override fun close() {
		inputStream.close()
	}

	@Throws(IOException::class)
	override fun available(): Int {
		return inputStream.available()
	}

	@Throws(IOException::class)
	override fun skip(n: Long): Long {
		val result = inputStream.skip(n)
		transferred += result
		bytesTransferred(transferred)
		return result
	}

	abstract fun bytesTransferred(transferred: Long)

	companion object {

		private const val EOF = -1
	}
}
